docs/custom-filters.adoc

```diff
@@ -45,23 +229,220 @@ Instances are specific to the channel between a single client and a single broke
+.----------------------------------------------------------------------------------------------------------------------.
+[0,0]: {"fill":"#99d","a2s:delref":1}
+....
+
+For example:
+
+[source,java]
+----
+public class ApiVersionsErrorFilter implements ApiVersionsRequestFilter {
+
+    @Override
+    public void onApiVersionsRequest(short apiVersion,
+                                     RequestHeaderData header,
+                                     ApiVersionsRequestData request,
+                                     KrpcFilterContext context) {
+        ApiVersionsResponseData response = new ApiVersionsResponseData(); //<1>
+        response.setErrorCode(Errors.UNKNOWN_SERVER_ERROR.code());
+        context.forwardResponse(response);//<2>
+    }
+}
+----
+<1> Create a new instance of the corresponding response data and populate it. Note you may need to use the `apiVersion`
+to check which fields can be set at this request's API version.
+<2> We forward the response, sending it towards the client, invoking Filters downstream of this one. Again, `forwardResponse`
+must be called during the invocation of the `onApiVersionsRequest` method.
+
+This will respond to all Api Versions requests with an error response without forwarding any of those requests to the Cluster.
+
+==== Sending out-of-band requests to the Cluster
+
+Filters can also send arbitrary new messages towards the Cluster that will be invisible to the Client. For example, we could send a
+request towards the Cluster for topic metadata while handling a FetchRequest.
+
+For example:
+
+[source,java]
+----
+public class FetchFilter implements FetchRequestFilter{
+    public static final short METADATA_VERSION_SUPPORTING_TOPIC_IDS = (short) 12;
+
+    @Override
+    public void onFetchRequest(short apiVersion,
+                               RequestHeaderData header,
+                               FetchRequestData request,
+                               KrpcFilterContext context) {
+        MetadataRequestData metadataRequest = new MetadataRequestData(); //<1>
+        var topic = new MetadataRequestData.MetadataRequestTopic();
+        topic.setTopicId(Uuid.randomUuid());
+        metadataRequest.topics().add(topic);
+        short version = METADATA_VERSION_SUPPORTING_TOPIC_IDS;
+        CompletionStage<MetadataResponseData> stage = context.sendRequest(version, request);//<2>
+        stage.thenAccept(response -> doSomethingWithMetadata(response, version));//<3>
+        context.forwardRequest(header, request);//<4>
+    }
+}
+----
+<1> We construct a new request object
+<2> We send the request towards the Cluster, specifying the api version to use and obtaining a CompletionStage for the
+response.
+<3> When the CompletionStage is completed, we do something with the metadata (populate a cache for example). Note that this
+stage is completed on the same thread as the rest of the Filter work, so we can safely mutate Filter members without
+synchronising.
```

Comment:
Perhaps its worth bringing this note out earlier in the filter descriptions as the fact that they are thread safe and don't require synchronisation is quite an important point for perfromance.

Reply:
There was an existing section `Handling state` covering it, but I've moved it up higher so it comes before the `Filter Patterns` section and linked off to the `sending out of band requests` section.